src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/VsTargetFrameworkInfo.cs

```diff
@@ -11,33 +11,67 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.PackageRestore;
 
-    public IVsReferenceItems PackageReferences => _packageReferences ??= new VsReferenceItems(_targetFrameworkInfo.PackageReferences);
+    public IReadOnlyDictionary<string, IReadOnlyList<IVsReferenceItem2>> Items
+    {
+        get
+        {
+            if (_items is null)
+            {
+                Span<int> lengths = stackalloc int[]
+                {
+                    _targetFrameworkInfo.FrameworkReferences.Length,
+                    _targetFrameworkInfo.NuGetAuditSuppress.Length,
+                    _targetFrameworkInfo.PackageDownloads.Length,
+                    _targetFrameworkInfo.PackageReferences.Length,
+                    _targetFrameworkInfo.CentralPackageVersions.Length,
+                    _targetFrameworkInfo.ProjectReferences.Length,
+                };
+                int capacity = lengths[0];
+                for (int i = 1; i < lengths.Length; i++)
+                {
+                    if (lengths[i] > capacity)
+                        capacity = lengths[i];
+                }
 
-    public IVsProjectProperties Properties => _properties ??= new VsProjectProperties(_targetFrameworkInfo.Properties);
+                ImmutableArray<IVsReferenceItem2>.Builder builder = ImmutableArray.CreateBuilder<IVsReferenceItem2>(capacity);
+                _items = ImmutableDictionary.CreateRange(
+                [
+                    new KeyValuePair<string, IReadOnlyList<IVsReferenceItem2>>("FrameworkReference", CreateImmutableVsReferenceItemList(_targetFrameworkInfo.FrameworkReferences, builder)),
+                    new KeyValuePair<string, IReadOnlyList<IVsReferenceItem2>>("NuGetAuditSuppress", CreateImmutableVsReferenceItemList(_targetFrameworkInfo.NuGetAuditSuppress, builder)),
+                    new KeyValuePair<string, IReadOnlyList<IVsReferenceItem2>>("PackageDownload", CreateImmutableVsReferenceItemList(_targetFrameworkInfo.PackageDownloads, builder)),
+                    new KeyValuePair<string, IReadOnlyList<IVsReferenceItem2>>("PackageReference", CreateImmutableVsReferenceItemList(_targetFrameworkInfo.PackageReferences, builder)),
+                    new KeyValuePair<string, IReadOnlyList<IVsReferenceItem2>>("PackageVersion", CreateImmutableVsReferenceItemList(_targetFrameworkInfo.CentralPackageVersions, builder)),
+                    new KeyValuePair<string, IReadOnlyList<IVsReferenceItem2>>("ProjectReference", CreateImmutableVsReferenceItemList(_targetFrameworkInfo.ProjectReferences, builder)),
+                ]);
+            }
+            return _items;
+        }
+    }
 
-    public IVsReferenceItems CentralPackageVersions => _centralPackageVersions ??= new VsReferenceItems(_targetFrameworkInfo.CentralPackageVersions);
+    private static IReadOnlyList<IVsReferenceItem2> CreateImmutableVsReferenceItemList(ImmutableArray<ReferenceItem> referenceItems, ImmutableArray<IVsReferenceItem2>.Builder builder)
+    {
+        builder.Clear();
```

Comment:
Rather than re-using a builder here, create a new one each time with exactly the right capacity, then use `MoveToImmutableArray` instead of `ToImmutableArray`. That way you'll create N arrays, not N + 1.

That approach will avoid allocating any more arrays than required. You can also then remove the code to compute the capacity above, and not pass the builder around. Better perf and clearer code.